File: export_pdf.py
import subprocess
import sys
import os

def check_and_install(package, import_name=None):
    import_name = import_name or package
    try:
        __import__(import_name)
        return True
    except ImportError:
        print(f"Cài đặt {package}...")
        subprocess.check_call([sys.executable, "-m", "pip", "install", package, "-q"])
        return True

check_and_install("weasyprint")

from weasyprint import HTML, CSS
from weasyprint.text.fonts import FontConfiguration

html_path = os.path.join(os.path.dirname(__file__), "quantum_report.html")
pdf_path  = os.path.join(os.path.dirname(__file__), "quantum_report.pdf")

font_config = FontConfiguration()

print("Đang xuất PDF...")
HTML(filename=html_path).write_pdf(
    pdf_path,
    font_config=font_config,
    presentational_hints=True,
)
print(f"Xuất thành công: {pdf_path}")

File: export_pdf.mjs
import puppeteer from 'puppeteer';
import { fileURLToPath } from 'url';
import { dirname, join } from 'path';

const __dirname = dirname(fileURLToPath(import.meta.url));
const htmlPath = join(__dirname, 'quantum_report.html');
const pdfPath  = join(__dirname, 'quantum_report.pdf');

console.log('Khởi động trình duyệt...');
const browser = await puppeteer.launch({ headless: true });
const page = await browser.newPage();

await page.goto(`file:///${htmlPath.replace(/\\/g, '/')}`, { waitUntil: 'networkidle0' });

// Đợi font load
await page.evaluate(() => document.fonts.ready);

console.log('Xuất PDF...');
await page.pdf({
  path: pdfPath,
  format: 'A4',
  printBackground: true,
  margin: { top: 0, right: 0, bottom: 0, left: 0 },
  preferCSSPageSize: true,
});

await browser.close();
console.log(`Xuất thành công: ${pdfPath}`);
